src/extensions/yfm/GPT/gptExtension/view.tsx

```diff
@@ -0,0 +1,322 @@
+        tr.setSelection(TextSelection.create(tr.doc, tr.mapping.map(to)));
+
+        this._view.dispatch(tr);
+
+        setTimeout(() => {
+            this._view.focus();
+        }, 0);
+
+        this._params.onApplyResult?.(answer);
+
+        this._resetState();
+    };
+
+    private _onClose = () => {
+        const deco = this._getCurrentDecoration();
+
+        if (!deco) {
+            return;
+        }
+
+        const tr = this._view.state.tr;
+        const meta: GptWidgetMeta = {action: 'hide'};
+
+        tr.setSelection(TextSelection.create(tr.doc, deco.from, deco.to));
+
+        tr.setMeta(pluginKey, meta);
+
+        this._view.dispatch(tr);
+
+        setTimeout(() => {
+            this._view.focus();
+        }, 0);
+
+        this._params.onClose?.();
+
+        this._resetState();
+    };
+
+    private _getContentOfDecoration(): string | undefined {
+        const deco = this._getCurrentDecoration();
+
+        if (!deco) return undefined;
+
+        const {from, to} = deco;
+
+        try {
+            const fragment = this._view.state.doc.slice(from, to, true).content;
+            const yfmMarkup = this._serializer.serialize(fragment);
+
+            return yfmMarkup;
+        } catch (error) {
+            console.error(error);
+
+            return this._view.state.doc.textBetween(from, to, '\n', '');
+        }
+    }
+
+    private _getCurrentDecoration() {
+        return this._getPluginState()?.find(
+            undefined,
+            undefined,
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-return
+            (spec) => spec[WIDGET_DECO_SPEC_FLAG],
+        )[0];
+    }
+
+    private _getPluginState() {
+        return pluginKey.getState(this._view.state);
+    }
+}
+
+type WidgetProps = Pick<PopupProps, 'anchorRef' | 'container'> & {
+    markup: string;
+    onClose: () => void;
+    confirmOpen: boolean;
+    onConfirmOk: () => void;
+    onConfirmCancel: () => void;
+} & GptDialogProps;
+
+function Widget({
+    markup,
+    anchorRef,
+    answerRender,
+    promptPresets,
+    disablePromptPresets,
+    customPromptPlaceholder,
+    disabledPromptPlaceholder,
+    onCustomPromptApply,
+    onApplyResult,
+    onPromptPresetClick,
+    onTryAgain,
+    onLike,
+    onDislike,
+    onClose,
+    onUpdate,
+    container,
+    gptAlertProps,
+}: WidgetProps) {
+    useEffect(() => {
+        // rerender the popup
+        window.dispatchEvent(new CustomEvent('scroll'));
+    }, [anchorRef]);
+
+    useMount(() => {
+        if (anchorRef?.current && 'scrollIntoView' in anchorRef.current) {
+            anchorRef.current.scrollIntoView({
+                behavior: 'smooth',
+                block: 'center',
+            });
+        }
+    });
+
+    const handleUpdate = useCallback(
+        (result?: CommonAnswer) => {
+            onUpdate?.(result);
+            requestAnimationFrame(() => {
```

Comment:
It may make sense to cancel the requestAnimationFrame when unmounting a component

https://github.com/gravity-ui/markdown-editor/blob/main/src/react-utils/useNodeResizing.ts#L99